src/idom/utils.py

```diff
@@ -63,80 +63,109 @@ def html_to_vdom(source: str, *transforms: _ModelTransform) -> Dict[str, Any]:
             dictionary which will be replaced by ``new``. For example, you could use a
             transform function to add highlighting to a ``<code/>`` block.
     """
-    parser = HtmlParser()
-    parser.feed(source)
-    root = parser.model()
-    to_visit = [root]
-    while to_visit:
-        node = to_visit.pop(0)
-        if isinstance(node, dict) and "children" in node:
-            transformed = []
-            for child in node["children"]:
-                if isinstance(child, dict):
-                    for t in transforms:
-                        child = t(child)
-                if child is not None:
-                    transformed.append(child)
-                    to_visit.append(child)
-            node["children"] = transformed
-            if "attributes" in node and not node["attributes"]:
-                del node["attributes"]
-            if "children" in node and not node["children"]:
-                del node["children"]
-    return root
-
-
-class HtmlParser(_HTMLParser):
-    """HTML to VDOM parser
-
-    Example:
-
-        .. code-block::
-
-            parser = HtmlParser()
-
-            parser.feed(an_html_string)
-            parser.feed(another_html_string)
-            ...
-
-            vdom = parser.model()
-    """
-
-    def model(self) -> Dict[str, Any]:
-        """Get the current state of parsed VDOM model"""
-        return self._node_stack[0]
 
-    def feed(self, data: str) -> None:
-        """Feed in HTML that will update the :meth:`HtmlParser.model`"""
-        self._node_stack.append(self._make_vdom("div", {}))
-        super().feed(data)
-
-    def reset(self) -> None:
-        """Reset the state of the parser"""
-        self._node_stack: List[Dict[str, Any]] = []
-        super().reset()
-
-    def handle_starttag(self, tag: str, attrs: List[Tuple[str, Optional[str]]]) -> None:
-        new = self._make_vdom(tag, dict(attrs))
-        current = self._node_stack[-1]
-        current["children"].append(new)
-        self._node_stack.append(new)
-
-    def handle_endtag(self, tag: str) -> None:
-        del self._node_stack[-1]
-
-    def handle_data(self, data: str) -> None:
-        self._node_stack[-1]["children"].append(data)
+    if not isinstance(html, str):
+        raise TypeError("html_to_vdom expects a string!")
+
+    return _HtmlToVdom(*transforms).convert(html)
+
+
+class _HtmlToVdom:
```

Comment:
It seems like this could be implemented with pure functions instead of a class. If so, let's do that. The only reason for the class before was that we had to subclass `HTMLParser`.

Reply:
I did have it implemented with pure functions before, but it I converted to a class based approach since it seemed easier to read/organize.

I'm not opposed to going back to functions though, but the `utils.py` file would then be littered with one-use helper functions that exist just for internal `html_to_vdom` use.

Reply:
I converted it back to a functional approach. Let me know which you prefer.